### PR TITLE
EZP-27445: Add actions for draft and archived versions

### PR DESCRIPTION
**Pull request by maxbaldanza**
- [x] Add delete/new/edit actions to draft and archived versions
- [x] Handle submission of deleting a draft version

https://jira.ez.no/browse/EZP-27566
screen shot 2017-06-30 at 09 59 46

### CODE REVIEW

**Comment by bdunogier** [comment]
We definitely need to use a custom controller for this. It is _really_ not a content _view_ thing :)

File: src/bundle/Resources/views/content/tabs/versions.html.twig
Comment (dpobel) [inline src/bundle/Resources/views/content/tabs/versions.html.twig]: the class should be `ez-list-toolbar-tools`

File: src/bundle/Resources/views/content/tabs/versions.html.twig
Comment (dpobel) [inline src/bundle/Resources/views/content/tabs/versions.html.twig]: why putting `ez-js-standard-form` class ?

File: src/bundle/Resources/views/content/tabs/versions.html.twig
Comment (dpobel) [inline src/bundle/Resources/views/content/tabs/versions.html.twig]: I guess (from your screenshot) `form_row` does not just generate the button, is there a way to only have the button without any extra markup ?

File: src/bundle/Resources/views/content/tabs/versions/table.html.twig
Comment (dpobel) [inline src/bundle/Resources/views/content/tabs/versions/table.html.twig]: that won't work (I know this comes from the UI guidelines...) If you really want to fill that column, for now just write something like:

```html
<td>
    <a href="#" class="ez-button" data-disabled>Edit</a>
</td>
```

**Review by dpobel — CHANGES_REQUESTED** [review]

File: src/bundle/Resources/views/content/tabs/versions.html.twig
Comment (maxbaldanza) [reply src/bundle/Resources/views/content/tabs/versions.html.twig]: There looks to be some JS that listens to form submit and makes this an ajax request. To have it post in the normal way I've added this class: https://github.com/ezsystems/hybrid-platform-ui-core-components/blob/master/js/ez-platform-ui-app.js#L53

File: src/bundle/Resources/views/content/tabs/versions.html.twig
Comment (maxbaldanza) [reply src/bundle/Resources/views/content/tabs/versions.html.twig]: It wraps the submit in a div: `<div><button type="submit" id="archived_actions_new_draft" name="archived_actions[new_draft]" class="ez-button ez-button-secondary">New draft based on selected version</button></div>`
Does this outer div need removing?

File: src/bundle/Resources/views/content/tabs/versions/table.html.twig
Comment (maxbaldanza) [reply src/bundle/Resources/views/content/tabs/versions/table.html.twig]: I assumed some CSS would be added at a later point for these to be displayed and it would be fine to leave like this for now but I'll update to your suggested value

File: src/bundle/Resources/views/content/tabs/versions.html.twig
Comment (maxbaldanza) [reply src/bundle/Resources/views/content/tabs/versions.html.twig]: Symfony forms are quite extensible so it can be removed if needed: https://symfony.com/doc/current/form/form_customization.html
I presume it will need removing from all buttons throughout the repo?

File: src/bundle/Resources/views/content/tabs/versions.html.twig
Comment (dpobel) [reply src/bundle/Resources/views/content/tabs/versions.html.twig]: yeah that's the whole point of the app :) (btw that's documented in https://github.com/ezsystems/hybrid-platform-ui/blob/master/docs/howto/prevent_navigation_enhancement.md)
As you wrote, if you don't put it, the form is submitted in AJAX which should not be an issue but maybe you experienced some with that ?

File: src/bundle/Resources/views/content/tabs/versions.html.twig
Comment (dpobel) [reply src/bundle/Resources/views/content/tabs/versions.html.twig]: yes please, maybe `{{ form_widget(archivedActionsForm.new_draft) }}` would do the trick ?

**Comment by bdunogier** [comment]
As said during the daily, the toggling of checkboxes is postponed to a later story.

The same feature is required by multiple tabs, and can be applied to all of them at once when we have it.

File: src/bundle/Resources/views/content/tabs/versions.html.twig
Comment (SylvainGuittard) [inline src/bundle/Resources/views/content/tabs/versions.html.twig]: It should be `desc('Delete')` instead of `desc('Delete selected draft')` (_cf._ mockup)

File: src/bundle/Resources/views/content/tabs/versions.html.twig
Comment (SylvainGuittard) [inline src/bundle/Resources/views/content/tabs/versions.html.twig]: It should be `desc('Delete')` instead of `desc('Remove selected archived version(s)')` (_cf._ mockup)

File: src/bundle/Resources/views/content/tabs/versions.html.twig
Comment (maxbaldanza) [inline src/bundle/Resources/views/content/tabs/versions.html.twig]: No `form_widget` has the same output. As far as I'm aware we'd have to customise that template. I'll put something together 😄

File: src/bundle/Resources/views/content/tabs/versions.html.twig
Comment (maxbaldanza) [reply src/bundle/Resources/views/content/tabs/versions.html.twig]: Is this the right class? I've added this class and the buttons show on top of each other as per my screenshot but when I remove the class they display side by side?

Comment (maxbaldanza) [reply src/bundle/Resources/views/content/tabs/versions.html.twig]: Ignore my previous comment 😄 I think the issue was my browser window was too small and therefore it was stacking them on top of each other.

File: src/bundle/Form/Versions/Data.php
Comment (bdunogier) [inline src/bundle/Form/Versions/Data.php]: Short array syntax.

File: src/bundle/Controller/ContentViewController.php
Comment (bdunogier) [inline src/bundle/Controller/ContentViewController.php:76]: I think it would make sense to move the controller action to the `VersionController`.

File: src/lib/Form/UiFormFactory.php
Comment (bdunogier) [inline src/lib/Form/UiFormFactory.php]: Would it make sense to provide the initialization data to the factory methods ? In that case, the array of versions.

The benefit of the FormFactory is quite limited in the current fashion, is it ?

File: src/bundle/Controller/ContentViewController.php
Comment (bdunogier) [inline src/bundle/Controller/ContentViewController.php]: `$contentInfo` isn't an argument of `mapToForm()`.

File: src/bundle/Controller/ContentViewController.php
Comment (bdunogier) [inline src/bundle/Controller/ContentViewController.php]: `$contentInfo` isn't an argument of `mapToForm()`.

**Review by bdunogier — CHANGES_REQUESTED** [review]

File: src/bundle/Controller/ContentViewController.php
Comment (maxbaldanza) [reply src/bundle/Controller/ContentViewController.php:76]: I think it would be a good idea but this PR is already quite big and is holding up some of the other PR's. We would also need to refactor some of it because it uses this `$this->getTranslations` and `$this->loadUser`. Can it be done afterwards in a separate PR?

File: src/lib/Repository/VersionService.php
Comment (andrerom) [inline src/lib/Repository/VersionService.php]: Like in #39 this should be named `UiVersionService` now right?

Comment (maxbaldanza) [reply src/lib/Repository/VersionService.php]: I'm not sure, I personally don't like the idea of prefixing classes unless they are needed. The reason the others are prefixed is to not confuse them with the ones in the kernel. As a `Version` service does not exist it doesn't technically need it. Do you think it's better to prefix them all for consistency or just the ones that are needed? 
I suppose if we prefix them all then it would avoid conflicts if one was added in the kernel at a later date?

**Review by mcurran-inviqa — APPROVED** [review]

File: src/lib/Repository/VersionService.php
Comment (andrerom) [reply src/lib/Repository/VersionService.php]: > I suppose if we prefix them all then it would avoid conflicts if one was added in the kernel at a later date?

ok with me

Comment (maxbaldanza) [reply src/lib/Repository/VersionService.php]: 👍 Already updated as I think that makes the most sense

File: src/bundle/Controller/ContentViewController.php
Comment (andrerom) [reply src/bundle/Controller/ContentViewController.php:76]: It can be done afterwards, but please then do that tomorrow in sync with @bdunogier

**Review by andrerom — APPROVED** [review]

**Comment by andrerom** [comment]
Merged to unblock the other PR's, @maxbaldanza please take care about feedback from @bdunogier tomorrow like the action move to VersionController.

**Pull request merged**
